File: 01-Arrays-List/twosum.py
#!/usr/bin/python3
#-*-coding:utf-8-*-
# @Time:    2020/9/3 7:11
# @File:    twosum.py

"""
1. 两数之和
https://leetcode-cn.com/problems/two-sum/
"""

from typing import List

class Solution(object):
    def twoSum(self, nums, target):
        dic = {}
        for i, n in enumerate(nums):
            if n in dic:
                return [dic[n], i]
            dic[target-n] = i

    def twoSum2(self, nums: List[int], target: int) -> List[int]:
        index = [[] for _ in range(2)]
        
        for i in range(len(nums)):
            for j in range(i+1,len(nums)):
                if nums[i] + nums[j] == target:
                    index[0] = i
                    index[1] = j
                    return index

if __name__ == "__main__":
    num = Solution()
    result = num.twoSum([1, 11, 7, -2,15],9)
    print(result)

File: 01-Arrays-List/twoSum.java
import java.util.Arrays;
import java.util.HashMap;
import java.util.Map;

public class twoSum {

    public static void main(String[] args){
        int[] nums = {1, 7, 11, -2,15};
        // int[] res = Solution(nums,9);
        int[] res2 = twoSum2(nums,9);
        System.out.println(
                Arrays.toString(res2));

    }

    public static int[] Solution(int[] nums, int target) {
        int[] index = new int[2];
        for (int i = 0; i < nums.length; i++) {
            for (int j = i + 1; j < nums.length; j++) {
                if (nums[i] + nums[j] == target) {
                    index[0] = i;
                    index[1] = j;
                    return index;
                }
            }
        }
        return new int[0];
    }

    public static int[] twoSum2(int[] nums, int target) {
        Map<Integer, Integer> hashtable = new HashMap<Integer, Integer>();
        for (int i = 0; i < nums.length; ++i) {
            if (hashtable.containsKey(nums[i])) {
                return new int[]{hashtable.get(nums[i]), i};
            }
            hashtable.put(target - nums[i], i);
        }
        return new int[0];
    }
}